jvmhs/src/Jvmhs/Analysis/Reduce.hs

```diff
@@ -11,4 +12,85 @@ This module defines the a class hierarchy analysis.
 module Jvmhs.Analysis.Reduce
   where
 
-import Control.Lens
+import            Control.Lens
+import            Jvmhs
+import            Data.Foldable
+
+import qualified Data.Set                             as S
+import qualified Data.Map                             as M
+
+
+findUnusedInterfaces ::
+    (Foldable t, MonadClassPool m)
+  => t ClassName
+  -- ^ A world to search for interfaces
+  -> m (M.Map ClassName (S.Set ClassName))
+  -- ^ A map from interfaces to the "correct" replacement of those interfaces.
+findUnusedInterfaces clsNames = do
+  clsLst <- traverse loadClass (toList clsNames)
+  let unusedInterfaces = S.difference
+        (findInterfaces clsLst)
+        (findUsedClasses clsLst)
+  unusedICls <- traverse loadClass (toList unusedInterfaces)
+  return $ inlineReplaceMap $
+    M.fromList
+      (map (\i -> (i^.className, S.fromList $ i^.classInterfaces)) unusedICls)
+
+
+inlineInterfaces ::
+  M.Map ClassName (S.Set ClassName)
+  -- ^ Collection of interfaces to inline
+  -> Class
+  -> Class
+  -- ^ If class has interface, replace it with the replacement of that interface
+inlineInterfaces replaceMap cls =
+  let oldInterface = cls^.classInterfaces
+  in cls & classInterfaces .~
+    foldMap (\i -> maybe [i] toList (replaceMap M.!? i)) oldInterface
+
+inlineReplaceMap ::
+     Ord a
+  => M.Map a (S.Set a)
+  -> M.Map a (S.Set a)
+inlineReplaceMap m =
+  if S.unions (M.elems m) `S.intersection` M.keysSet m == S.empty
+  then m
+  else inlineReplaceMap $
+         M.map inlineKey m
+    where inlineKey = S.fromList.foldMap
```

Comment:
You reuse this function from line 49, maybe extract it to a function?
Also there is a lot of conversion between lists and sets :)